File: CPP_OmniSoc/src/Socket_Serial.cpp
#include "Socket_Serial.h"

#include <chrono>
#include <iostream>

Socket_Serial::Socket_Serial(const std::string& _IP_Address, const std::string& _port, bool _isServer, bool _asyncronousFlag)
    : io_context_(), socket_(io_context_) {
    
    asyncronousFlag = _asyncronousFlag;
    IP_Address = _IP_Address;
    port = _port;
    isServer = _isServer;
}

void Socket_Serial::synchronousUpdate()
{
    if (asyncronousFlag) { return; } //no need to call syncronousUpdate in async mode
    doConnection();
    doSerial();
}

Socket_Serial::~Socket_Serial() {
    disconnect();
}

void Socket_Serial::connect(bool blocking_flag, bool auto_reconnect, int _period_ms) {
    if (!asyncronousFlag) { return; }

    period_ms = _period_ms;
    autoReconnect = auto_reconnect;

    connection_thread_ = std::thread(&Socket_Serial::connectionThread, this);

    if (blocking_flag)
    {
        while (!connectedFlag)
        {
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }
    }
}

void Socket_Serial::disconnect() {
    try
    {
        if (acceptor_ != nullptr && acceptor_->is_open())
        { acceptor_->close(); }
    }
    catch(...){}

    try
    {
        std::cout << "Connection Closed" << std::endl;
        autoReconnect = false;
        killFlag = true;

        if (connection_thread_.joinable())
        { connection_thread_.join(); }

        if (serial_thread_.joinable())
        { serial_thread_.join(); }

        closeSocket();
    }
    catch (const std::exception& e) {
        if (!suppressCatchPrints) { std::cerr << "Exception caught: " << e.what() << std::endl; }
    }
    catch (...) {
        if (!suppressCatchPrints) { std::cerr << "Unknown exception caught." << std::endl; }
    }
}

void Socket_Serial::send(const std::string& msg) {
    std::lock_guard<std::mutex> lock(out_buffer_mutex_);
    outgoing_buffer_.push_back(msg);
}

std::vector<std::string> Socket_Serial::receive(int count) {
    std::vector<std::string> messages;
    std::unique_lock<std::mutex> lock(in_buffer_mutex_);

    if (count == -1) {
        messages = std::move(incoming_buffer_);
        incoming_buffer_.clear();
    }
    else {
        int num_messages = std::min(count, static_cast<int>(incoming_buffer_.size()));
        messages.insert(messages.end(), incoming_buffer_.begin(), incoming_buffer_.begin() + num_messages);
        incoming_buffer_.erase(incoming_buffer_.begin(), incoming_buffer_.begin() + num_messages);
    }

    return messages;
}

bool Socket_Serial::isConnected() {
    return connectedFlag;
}

void Socket_Serial::clearInBuffer() {
    std::lock_guard<std::mutex> lock(in_buffer_mutex_);
    incoming_buffer_.clear();
}

void Socket_Serial::clearOutBuffer() {
    std::lock_guard<std::mutex> lock(out_buffer_mutex_);
    outgoing_buffer_.clear();
}

void Socket_Serial::flushSocket() {
    // No need to flush for TCP/IP sockets
}

void Socket_Serial::connectionThread() {
    if (!asyncronousFlag) { return; }

    bool connectionOneShot = true;

    while (!killFlag && (autoReconnect || connectionOneShot)) {
        if (!connectedFlag)
        {
            if (serial_thread_.joinable())
            { serial_thread_.join(); }

            doConnection();

            if (socket_.is_open()) 
            {
                serial_thread_ = std::thread(&Socket_Serial::serialThread, this);
                connectionOneShot = false;
            }
        }
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
}

void Socket_Serial::doConnection()
{
    try {
        if (!connectedFlag) {

            boost::asio::ip::tcp::resolver resolver(io_context_);
            //boost::asio::ip::tcp::resolver::query query(boost::asio::ip::tcp::v4(), IP_Address, port);
            //boost::asio::ip::tcp::resolver::iterator endpoint_iterator = resolver.resolve(query);
            auto endpoints = resolver.resolve(IP_Address, port);

            if (isServer) {
                acceptor_ = std::make_shared < boost::asio::ip::tcp::acceptor>(io_context_, *endpoints.begin());
                bool acceptError = false;
                acceptor_->accept(socket_);
            }
            else {
                boost::asio::connect(socket_, endpoints);
            }

            if (socket_.is_open()) {
                socket_.non_blocking(true);
                std::cout << "socket connected" << std::endl;
                missedHeartbeats = -5; //slight grace period for initial connection.
                connectedFlag = true;
            }
        }
    }
    catch (const std::exception& e) {
        // Catch and print standard exceptions
        if (!suppressCatchPrints) { std::cout << "Connection Exception: " << e.what() << std::endl; }
        //closeSocket();
    }
    catch (...) {
        // Catch and print any other exception
        if (!suppressCatchPrints) { std::cout << "Unknown exception caught in connection." << std::endl; }
    }
}

void Socket_Serial::closeSocket()
{
    if (connectedFlag)
    {
        std::cout << "Connection Dropped" << std::endl;
    }

    boost::system::error_code ec;

    if(socket_.is_open())
    {
        socket_.shutdown(boost::asio::ip::tcp::socket::shutdown_both, ec);
        if (ec && !suppressCatchPrints) {
            std::cerr << "Failed to shutdown socket: " << ec.message() << std::endl;
        }
        socket_.close(ec);
        if (ec && !suppressCatchPrints) {
            std::cerr << "Failed to close socket: " << ec.message() << std::endl;
        }
        socket_.release(ec);
        if (ec && !suppressCatchPrints) {
            std::cerr << "Failed to release socket: " << ec.message() << std::endl;
        }
    }

    if(acceptor_ != nullptr)
    {
        acceptor_->close(ec);
        if (ec && !suppressCatchPrints) {
            std::cerr << "Failed to close acceptor: " << ec.message() << std::endl;
        }

        acceptor_->release(ec);
        if (ec && !suppressCatchPrints) {
            std::cerr << "Failed to release acceptor: " << ec.message() << std::endl;
        }
    }

    connectedFlag = false;
}

void Socket_Serial::serialThread() {
    if (!asyncronousFlag) { return; }

    while (!killFlag && connectedFlag) {
        doSerial();
        std::this_thread::sleep_for(std::chrono::milliseconds(period_ms));
    }
}

void Socket_Serial::doSerial()
{
    sendMessages();
    readMessages();
}

void Socket_Serial::sendMessages() {
    if (!connectedFlag) { return; }

    std::lock_guard<std::mutex> lock(out_buffer_mutex_);

    try
    {
        if (!outgoing_buffer_.empty()) {
            for (const auto& msg : outgoing_buffer_) {
                boost::asio::write(socket_, boost::asio::buffer(msg));
                boost::asio::write(socket_, boost::asio::buffer(msgDelimiter));
            }
            outgoing_buffer_.clear();
        }
        else {
            boost::asio::write(socket_, boost::asio::buffer(msgDelimiter));
        }
    }
    catch (...) {
        if (!suppressCatchPrints) { std::cerr << "Write error: " << std::endl; }
        closeSocket();
    }
}

void Socket_Serial::readMessages() {
    if (!connectedFlag) { return; }

    try
    {
        char buffer[1024];
        boost::system::error_code error;
        size_t bytes_read = socket_.read_some(boost::asio::buffer(buffer), error);

        if (error == boost::asio::error::eof || error == boost::asio::error::would_block) {
            missedHeartbeats++;
            if (missedHeartbeats >= missedHeartbeatLimit) {
                std::cout << "Heartbeat kill: " << missedHeartbeats << std::endl;
                closeSocket();
            }
        }
        else if (!error) {
            if (bytes_read > 0) {
                std::string message(buffer, bytes_read);
                std::string tempRemainder;
                std::vector<std::string> msgs = splitMessage(message, msgDelimiter, tempRemainder);
                for (int i = 0; i < msgs.size(); i++)
                {
                    std::string isolatedMessage = "";
                    if (i == 0 && inMessageRemainder.size()>0)
                    { isolatedMessage = inMessageRemainder + msgs[i]; }
                    else
                    { isolatedMessage = msgs[i]; }

                    if (isolatedMessage.size() > 0)
                    {
                        std::lock_guard<std::mutex> lock(in_buffer_mutex_);
                        incoming_buffer_.emplace_back(isolatedMessage);
                    }
                }
                if (msgs.size() > 0)
                { inMessageRemainder = tempRemainder; }
                else
                { inMessageRemainder = inMessageRemainder + tempRemainder; }
                
                missedHeartbeats = 0;
            }
        }
        else
        {
            closeSocket();
        }
    }
    catch (...) {
        if (!suppressCatchPrints) { std::cerr << "Read error: " << std::endl; }
        closeSocket();
    }
}

std::vector<std::string> Socket_Serial::splitMessage(const std::string& message, const std::string& delimiter) {
    std::string inMsg = message;
    std::vector<std::string> messages;
    size_t pos = 0;
    std::string token;

    while ((pos = inMsg.find(delimiter)) != std::string::npos) {
        token = inMsg.substr(0, pos);
        messages.push_back(token);
        inMsg.erase(0, pos + delimiter.length());
    }

    if (!inMsg.empty()) {
        messages.push_back(inMsg);
    }

    return messages;
}

File: Unity_Port/SocketSerial/Runtime/SocketSerial.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using UnityEngine;

/// <summary>
/// Unity C# port of Socket_Serial for TCP/IP communication
/// Supports both client and server modes with async or sync operation
/// </summary>
public class SocketSerial
{
    private TcpClient client;
    private TcpListener server;
    private NetworkStream stream;
    
    private Thread connectionThread;
    private Thread serialThread;
    
    private readonly object inBufferLock = new object();
    private readonly object outBufferLock = new object();
    
    private List<string> incomingBuffer = new List<string>();
    private List<string> outgoingBuffer = new List<string>();
    
    private bool asyncronousFlag = false;
    private bool autoReconnect = false;
    private int missedHeartbeats = 0;
    private bool isServer = false;
    private bool connectedFlag = false;
    private bool killFlag = false;
    private string msgDelimiter = ";";
    private string inMessageRemainder = "";
    
    private string ipAddress;
    private int port;
    private int periodMs;
    
    // Public settings
    public bool SuppressCatchPrints { get; set; } = true;
    public int MissedHeartbeatLimit { get; set; } = 50;
    
    /// <summary>
    /// Create a new SocketSerial instance
    /// </summary>
    /// <param name="_ipAddress">IP address to connect to (client) or bind to (server)</param>
    /// <param name="_port">Port number</param>
    /// <param name="_isServer">True for server mode, false for client mode</param>
    /// <param name="_asyncronousFlag">True for async operation with threads, false for manual sync updates</param>
    public SocketSerial(string _ipAddress, int _port, bool _isServer, bool _asyncronousFlag = true)
    {
        asyncronousFlag = _asyncronousFlag;
        ipAddress = _ipAddress;
        port = _port;
        isServer = _isServer;
    }
    
    /// <summary>
    /// Connect to remote endpoint (async mode only)
    /// </summary>
    /// <param name="blockingFlag">If true, blocks until connected</param>
    /// <param name="autoReconnect">If true, automatically reconnect on disconnect</param>
    /// <param name="_periodMs">Update period in milliseconds for serial thread</param>
    public void Connect(bool blockingFlag = false, bool autoReconnect = false, int _periodMs = 10)
    {
        if (!asyncronousFlag) return;
        
        periodMs = _periodMs;
        this.autoReconnect = autoReconnect;
        
        connectionThread = new Thread(ConnectionThread);
        connectionThread.IsBackground = true;
        connectionThread.Start();
        
        if (blockingFlag)
        {
            while (!connectedFlag && !killFlag)
            {
                Thread.Sleep(50);
            }
        }
    }
    
    /// <summary>
    /// Disconnect and cleanup
    /// </summary>
    public void Disconnect()
    {
        try
        {
            if (server != null)
            {
                server.Stop();
            }
        }
        catch { }
        
        try
        {
            Debug.Log("Connection Closed");
            autoReconnect = false;
            killFlag = true;
            
            if (connectionThread != null && connectionThread.IsAlive)
            {
                connectionThread.Join(1000);
                if (connectionThread.IsAlive)
                    connectionThread.Abort();
            }
            
            if (serialThread != null && serialThread.IsAlive)
            {
                serialThread.Join(1000);
                if (serialThread.IsAlive)
                    serialThread.Abort();
            }
            
            CloseSocket();
        }
        catch (Exception e)
        {
            if (!SuppressCatchPrints)
                Debug.LogError($"Exception caught: {e.Message}");
        }
    }
    
    /// <summary>
    /// Send a message
    /// </summary>
    /// <param name="msg">Message to send (delimiter will be added automatically)</param>
    public void Send(string msg)
    {
        lock (outBufferLock)
        {
            outgoingBuffer.Add(msg);
        }
    }
    
    /// <summary>
    /// Receive messages from buffer
    /// </summary>
    /// <param name="count">Number of messages to retrieve, -1 for all</param>
    /// <returns>List of received messages</returns>
    public List<string> Receive(int count = -1)
    {
        List<string> messages = new List<string>();
        
        lock (inBufferLock)
        {
            if (count == -1)
            {
                messages.AddRange(incomingBuffer);
                incomingBuffer.Clear();
            }
            else
            {
                int numMessages = Mathf.Min(count, incomingBuffer.Count);
                messages.AddRange(incomingBuffer.GetRange(0, numMessages));
                incomingBuffer.RemoveRange(0, numMessages);
            }
        }
        
        return messages;
    }
    
    /// <summary>
    /// Check if currently connected
    /// </summary>
    public bool IsConnected()
    {
        return connectedFlag;
    }
    
    /// <summary>
    /// Clear incoming message buffer
    /// </summary>
    public void ClearInBuffer()
    {
        lock (inBufferLock)
        {
            incomingBuffer.Clear();
        }
    }
    
    /// <summary>
    /// Clear outgoing message buffer
    /// </summary>
    public void ClearOutBuffer()
    {
        lock (outBufferLock)
        {
            outgoingBuffer.Clear();
        }
    }
    
    /// <summary>
    /// Flush socket (no-op for TCP/IP)
    /// </summary>
    public void FlushSocket()
    {
        // No need to flush for TCP/IP sockets
    }
    
    /// <summary>
    /// Manual synchronous update for non-async mode
    /// Call this regularly in your Update() loop if asyncronousFlag is false
    /// If called too fast without delay, may crash due to heartbeat would_block errors
    /// </summary>
    public void SynchronousUpdate()
    {
        if (asyncronousFlag) return;
        
        DoConnection();
        DoSerial();
    }
    
    // Private methods
    
    private void ConnectionThread()
    {
        if (!asyncronousFlag) return;
        
        bool connectionOneShot = true;
        
        while (!killFlag && (autoReconnect || connectionOneShot))
        {
            if (!connectedFlag)
            {
                if (serialThread != null && serialThread.IsAlive)
                {
                    serialThread.Join(1000);
                    if (serialThread.IsAlive)
                        serialThread.Abort();
                }
                
                DoConnection();
                
                if (stream != null)
                {
                    serialThread = new Thread(SerialThread);
                    serialThread.IsBackground = true;
                    serialThread.Start();
                    connectionOneShot = false;
                }
            }
            
            Thread.Sleep(1000);
        }
    }
    
    private void DoConnection()
    {
        try
        {
            if (!connectedFlag)
            {
                if (isServer)
                {
                    if (server == null)
                    {
                        IPAddress ipAddr = IPAddress.Parse(ipAddress);
                        server = new TcpListener(ipAddr, port);
                        server.Start();
                        Debug.Log($"Server listening on {ipAddress}:{port}");
                    }
                    
                    // Non-blocking check for pending connections
                    if (server.Pending())
                    {
                        client = server.AcceptTcpClient();
                        stream = client.GetStream();
                        client.NoDelay = true;
                        stream.ReadTimeout = 100;
                        stream.WriteTimeout = 100;
                    }
                }
                else
                {
                    client = new TcpClient();
                    client.Connect(ipAddress, port);
                    stream = client.GetStream();
                    client.NoDelay = true;
                    stream.ReadTimeout = 100;
                    stream.WriteTimeout = 100;
                }
                
                if (stream != null)
                {
                    Debug.Log("Socket connected");
                    missedHeartbeats = -5; // Grace period for initial connection
                    connectedFlag = true;
                }
            }
        }
        catch (Exception e)
        {
            if (!SuppressCatchPrints)
                Debug.Log($"Connection Exception: {e.Message}");
        }
    }
    
    private void CloseSocket()
    {
        if (connectedFlag)
        {
            Debug.Log("Connection Dropped");
        }
        
        try
        {
            if (stream != null)
            {
                stream.Close();
                stream = null;
            }
            
            if (client != null)
            {
                client.Close();
                client = null;
            }
        }
        catch (Exception e)
        {
            if (!SuppressCatchPrints)
                Debug.LogError($"Failed to close socket: {e.Message}");
        }
        
        connectedFlag = false;
    }
    
    private void SerialThread()
    {
        if (!asyncronousFlag) return;
        
        while (!killFlag && connectedFlag)
        {
            DoSerial();
            Thread.Sleep(periodMs);
        }
    }
    
    private void DoSerial()
    {
        SendMessages();
        ReadMessages();
    }
    
    private void SendMessages()
    {
        if (!connectedFlag || stream == null) return;
        
        lock (outBufferLock)
        {
            try
            {
                if (outgoingBuffer.Count > 0)
                {
                    foreach (string msg in outgoingBuffer)
                    {
                        byte[] data = Encoding.UTF8.GetBytes(msg);
                        stream.Write(data, 0, data.Length);
                        
                        byte[] delimiter = Encoding.UTF8.GetBytes(msgDelimiter);
                        stream.Write(delimiter, 0, delimiter.Length);
                    }
                    outgoingBuffer.Clear();
                }
                else
                {
                    // Send heartbeat
                    byte[] delimiter = Encoding.UTF8.GetBytes(msgDelimiter);
                    stream.Write(delimiter, 0, delimiter.Length);
                }
            }
            catch (Exception e)
            {
                if (!SuppressCatchPrints)
                    Debug.LogError($"Write error: {e.Message}");
                CloseSocket();
            }
        }
    }
    
    private void ReadMessages()
    {
        if (!connectedFlag || stream == null) return;
        
        try
        {
            if (stream.DataAvailable)
            {
                byte[] buffer = new byte[1024];
                int bytesRead = stream.Read(buffer, 0, buffer.Length);
                
                if (bytesRead > 0)
                {
                    string message = Encoding.UTF8.GetString(buffer, 0, bytesRead);
                    string tempRemainder = "";
                    List<string> msgs = SplitMessage(message, msgDelimiter, out tempRemainder);
                    
                    for (int i = 0; i < msgs.Count; i++)
                    {
                        string isolatedMessage = "";
                        if (i == 0 && inMessageRemainder.Length > 0)
                        {
                            isolatedMessage = inMessageRemainder + msgs[i];
                        }
                        else
                        {
                            isolatedMessage = msgs[i];
                        }
                        
                        if (isolatedMessage.Length > 0)
                        {
                            lock (inBufferLock)
                            {
                                incomingBuffer.Add(isolatedMessage);
                            }
                        }
                    }
                    
                    if (msgs.Count > 0)
                    {
                        inMessageRemainder = tempRemainder;
                    }
                    else
                    {
                        inMessageRemainder = inMessageRemainder + tempRemainder;
                    }
                    
                    missedHeartbeats = 0;
                }
            }
            else
            {
                missedHeartbeats++;
                if (missedHeartbeats >= MissedHeartbeatLimit)
                {
                    Debug.Log($"Heartbeat kill: {missedHeartbeats}");
                    CloseSocket();
                }
            }
        }
        catch (Exception e)
        {
            if (!SuppressCatchPrints)
                Debug.LogError($"Read error: {e.Message}");
            CloseSocket();
        }
    }
    
    /// <summary>
    /// Split a message by delimiter
    /// </summary>
    public static List<string> SplitMessage(string message, string delimiter, out string remainder)
    {
        List<string> messages = new List<string>();
        string inMsg = message;
        
        int pos;
        while ((pos = inMsg.IndexOf(delimiter)) != -1)
        {
            string token = inMsg.Substring(0, pos);
            messages.Add(token);
            inMsg = inMsg.Substring(pos + delimiter.Length);
        }
        
        remainder = inMsg;
        return messages;
    }
}
